lib/ruby_lsp/indexer.rb

```diff
@@ -0,0 +1,68 @@
+# typed: true
+# frozen_string_literal: true
+
+class Indexer
+  VM_DEFINECLASS_TYPE_CLASS           = 0x00
+  VM_DEFINECLASS_TYPE_SINGLETON_CLASS = 0x01
+  VM_DEFINECLASS_TYPE_MODULE          = 0x02
+  VM_DEFINECLASS_FLAG_SCOPED          = 0x08
+  VM_DEFINECLASS_FLAG_HAS_SUPERCLASS  = 0x10
+
+  def initialize(root)
+    @root = root
+    @index = {
+      constant: (Hash.new { |h,k| h[k] = [] }),
+      method: (Hash.new { |h,k| h[k] = [] })
+    }
+  end
+
+  def run
+    Dir["#{@root}/**/*.rb"].each do |file|
+      index_file(file)
+    end
+  end
+
+  def locs_for_symbol(symbol, kind)
+    $stderr.puts "--- Looking for #{symbol} of kind #{kind}"
+    index.dig(kind, symbol)
+  end
+
+  private
+
+  attr_reader :root, :index
+
+  def index_file(file)
+    iseq = RubyVM::InstructionSequence.compile(File.read(file)).to_a
+    index_iseq(iseq) do |line, event, message|
+      index[event][message] << [File.absolute_path(file), line]
```

Comment:
Removing this block and instead having `index_iseq` either push directly into the hashes or call methods like `index_constant` and `index_method` would be a lot more performant.